### DISCUSSION TITLE
Unable to inject async DB connection via web::Dta<Pool<<>>>>

### DISCUSSION THREAD

**Discussion by ag039st** [weiznich/diesel_async#176] [Q&A]
## Setup
use actix_web::web::Data;
use actix_web::{web, HttpResponse, Responder};
use diesel_async::pooled_connection::deadpool::Pool;
use diesel_async::pooled_connection::AsyncDieselConnectionManager;
use diesel_async::AsyncPgConnection;

use crate::models::models::UserDto;

pub async fn create_user(
    user: web::Json<UserDto>,
    pool: Data<Pool<AsyncDieselConnectionManager<AsyncPgConnection>>>,
) -> impl Responder {
    // let mut connection: &mut Pool<AsyncDieselConnectionManager<AsyncPgConnection>>;
    // con.clone_into(connection);
    let mut conn = pool.get().await.unwrap();
    // create_user_and_profile(, &user.user, &user.profile).await
    HttpResponse::Ok().body("Ok")
}

use crate::controllers::user_controller::create_user;
use actix_web::dev::Server;
use actix_web::{middleware, web, App, HttpServer};
use diesel_async::pooled_connection::deadpool::Pool;
use diesel_async::pooled_connection::AsyncDieselConnectionManager;
use log::{error, info};
use std::env;
use std::env::var;
use std::error::Error;
use std::net::TcpListener;

/// Starts a web server on the specified listener.
///
/// # Arguments
///
/// * `listener` - A `TcpListener` instance used to listen for incoming connections.
///
/// # Returns
///
/// * `Result<Server, Box<dyn Error>>` - A Result indicating success or failure. If successful, returns a `Server` instance representing the running web server.
pub async fn start(listener: TcpListener) -> Result<Server, Box<dyn Error>> {
    let port = listener.local_addr().unwrap().port();
    let database_url = var("DATABASE_URL").unwrap_or_else(|e| panic!("an error occurred {e}"));
    let config = AsyncDieselConnectionManager::<diesel_async::AsyncPgConnection>::new(database_url);
    let pool = Pool::builder(config).build()?;

    // let db_pool = web::Data::new(pool);
    info!(">>> Starting web server on port: {} >>>", port);
    let server = HttpServer::new(|| {
        App::new()
            .route("/api/user", web::post().to(create_user))
            .app_data(pool.clone())
            .wrap(middleware::Logger::default())
    })
    .listen(listener)?
    .run();
    Ok(server)
}

### Versions
diesel-async = {version = "0.5.0", features = ["postgres","deadpool"]}
diesel = {version = "2.2.2", features = ["chrono", "postgres"]}

- **Rust:1.82.0 nightly
- **Diesel: 2.2.2
- **Diesel_async:0.5.0
- **Database:postgres
- **Operating System: Darwin BB-Mac-Pro.attlocal.net 24.0.0 Darwin Kernel Version 24.0.0: Sat Jul 13 00:54:59 PDT 2024; root:xnu-11215.0.165.0.4~50/RELEASE_ARM64_T6031 arm64

### Feature Flags

- **diesel: "chrono", "postgres"
- **diesel_async:"postgres","deadpool"

## Problem Description
error[E0277]: the trait bound `AsyncDieselConnectionManager<AsyncPgConnection>: DerefMut` is not satisfied
   --> src/controllers/user_controller.rs:11:16
    |
11  |     pool: Data<Pool<AsyncDieselConnectionManager<AsyncPgConnection>>>,
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `DerefMut` is not implemented for `AsyncDieselConnectionManager<AsyncPgConnection>`, which is required by `AsyncDieselConnectionManager<AsyncDieselConnectionManager<AsyncPgConnection>>: deadpool::managed::Manager`
    |
    = help: the trait `deadpool::managed::Manager` is implemented for `AsyncDieselConnectionManager<C>`
    = note: required for `AsyncDieselConnectionManager<AsyncPgConnection>` to implement `AsyncConnection`

### What are you trying to accomplish?
trying to inject Pool to app_Data() actix web 

### What is the expected output?
I expect the build to succeed and get connections from the pool 

### What is the actual output?
error[E0277]: the trait bound `AsyncDieselConnectionManager<AsyncPgConnection>: DerefMut` is not satisfied
   --> src/controllers/user_controller.rs:11:16
    |
11  |     pool: Data<Pool<AsyncDieselConnectionManager<AsyncPgConnection>>>,
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `DerefMut` is not implemented for `AsyncDieselConnectionManager<AsyncPgConnection>`, which is required by `AsyncDieselConnectionManager<AsyncDieselConnectionManager<AsyncPgConnection>>: deadpool::managed::Manager`
    |
    = help: the trait `deadpool::managed::Manager` is implemented for `AsyncDieselConnectionManager<C>`
    = note: required for `AsyncDieselConnectionManager<AsyncPgConnection>` to implement `AsyncConnection`

### Steps to reproduce

use actix_web::web::Data;
use actix_web::{web, HttpResponse, Responder};
use diesel_async::pooled_connection::deadpool::Pool;
use diesel_async::pooled_connection::AsyncDieselConnectionManager;
use diesel_async::AsyncPgConnection;

use crate::models::models::UserDto;

pub async fn create_user(
    user: web::Json<UserDto>,
    pool: Data<Pool<AsyncDieselConnectionManager<AsyncPgConnection>>>,
) -> impl Responder {
    // let mut connection: &mut Pool<AsyncDieselConnectionManager<AsyncPgConnection>>;
    // con.clone_into(connection);
    let mut conn = pool.get().await.unwrap();
    // create_user_and_profile(, &user.user, &user.profile).await
    HttpResponse::Ok().body("Ok")
}

## Checklist

- [x] I have already looked over the [issue tracker](https://github.com/weiznich/diesel_async/issues) for similar possible closed issues.
 
- [x] This issue can be reproduced on Rust's stable channel. (Your issue will be
  closed if this is not the case)
- [x] This issue can be reproduced without requiring a third party crate

**Comment by weiznich** [comment]
I converted this to a discussion as it's not a issue in diesel-async, but a support question. Please use the discussion forum for future similar questions as we use the issue tracker to track actual bugs. Future posts there will be handled as bug reports and closed as invalid without further explanations.

The error message already provides all relevant information to resolve this issue. In particular this part:

> ```
> which is required by AsyncDieselConnectionManager<AsyncDieselConnectionManager<AsyncPgConnection>>: deadpool::managed::Manager
> = help: the trait deadpool::managed::Manager is implemented for AsyncDieselConnectionManager<C>
> = note: required for AsyncDieselConnectionManager<AsyncPgConnection> to implement AsyncConnection
> ```

You just mixed up the types in your `Pool` definition. Instead of `Pool<AsyncDieselConnectionManager>` you likely want `Pool<AsyncPgConnection>`.

**Reply by ag039st** [reply]
thanks @weiznich  let me check that thanks

**Reply by ag039st** [reply]
do you have a example of insert with Pool<AsyncPgConnection>  ? - Im not succeeding with this , transactions seems not to exist and also get_result() do not loke Pool<AsyncPgConnection>

**Reply by weiznich** [reply]
An insert statement wouldn't look that different to the example given here: https://docs.rs/diesel-async/latest/diesel_async/pooled_connection/deadpool/index.html

The only notable difference is that you would replace the query.
